src/funmisc.c

```diff
@@ -1562,3 +1567,273 @@ FUNCTION(fun_benchmark)
 
   return;
 }
+
+#ifdef HAVE_LIBCURL
+static size_t
+req_write_callback(void *contents, size_t size, size_t nmemb, void *userp)
+{
+  struct urlreq *req = userp;
+  size_t realsize = size * nmemb;
+  sqlite3_str_append(req->body, contents, realsize);
+  if (sqlite3_str_length(req->body) >= BUFFER_LEN) {
+    /* Raise an error and abort request. */
+    req->too_big = 1;
+    return 0;
+  } else {
+    return realsize;
+  }
+}
+
+static int
+req_set_cloexec(void *clientp __attribute__((__unused__)), curl_socket_t fd,
+                curlsocktype purpose __attribute__((__unused__)))
+{
+  set_close_exec(fd);
+  return CURL_SOCKOPT_OK;
+}
+
+static void
+free_urlreq(struct urlreq *req)
+{
+  pe_regs_free(req->pe_regs);
+  if (req->body) {
+    sqlite3_str_reset(req->body);
+    sqlite3_str_finish(req->body);
+  }
+  mush_free(req->attrname, "urlreq.attrname");
+  curl_slist_free_all(req->header_slist);
+  mush_free(req, "urlreq");
+}
+#endif
+
+/* ARGSUSED */
+FUNCTION(fun_http)
+{
+  #ifdef HAVE_LIBCURL
+  ufun_attrib ufun;
+  extern int ncurl_queries;
+  CURLcode curl_status;
+  struct urlreq *req;
+  CURL *handle;
+  struct curl_slist *headers = NULL;
+  dbref thing;
+  char *s;
+  const char *userpass;
+  char tbuf[BUFFER_LEN];
+  // int queue_type = QUEUE_DEFAULT | (queue_entry->queue_type & QUEUE_EVENT);
+  bool post = false;
+  bool del = false;
+  bool put = false;
+
+  if (!Wizard(executor) && !has_power_by_name(executor, "Can_HTTP", NOTYPE)) {
+    safe_strl("#-1 Permission Denied.", 22, buff, bp);
+    return;
+  }
+  
+  if(nargs < 3) {
+    safe_str("#-1 FUNCTION EXPECTS AT LEAST 3 ARGUMENTS", buff, bp);
+    return;
+  }
+  
+  if(nargs > 4) {
+    safe_str("#-1 FUNCTION EXPECTS AT MOST 4 ARGUMENTS", buff, bp);
+    return;
+  }
+
+  if (!*args[2]) {
+    safe_strl("#-1 Missing URL.", 16, buff, bp);
+    return;
+  }
+
+  if (strcasecmp(args[1], "POST") == 0) {
+    post = true;
+  }
+
+  if (strcasecmp(args[1], "PUT") == 0) {
+    put = true;
+  }
+
+  if (strcasecmp(args[1], "DELETE") == 0) {
+    del = true;
+  }
+
+  if(!(post||put||del) && nargs > 3) {
+    safe_strl("#-1 A GET request does not support a body argument.", 51, buff, bp);
+    return;
+  }
+
+  mush_strncpy(tbuf, args[0], sizeof tbuf);
+  s = strchr(tbuf, '/');
+  if (!s) {
+    safe_strl("#-1 I need to know what attribute to trigger.", 45, buff, bp);
+    return;
+  }
+  *(s++) = '\0';
+  upcasestr(s);
+
+  thing = noisy_match_result(executor, tbuf, NOTYPE, MAT_EVERYTHING);
+
+  if (thing == NOTHING) {
+    return;
+  }
+
+  if (!controls(executor, thing)) {
+    safe_strl("#-1 PERMISSION DENIED", 21, buff, bp);
+    return;
+  }
+  
+  if (!fetch_ufun_attrib(args[0], executor, &ufun, UFUN_DEFAULT))
+  {
+    safe_strl("#-1 Invalid attribute to call", 29, buff, bp);
+    return;
+  }
+
+  req = mush_malloc(sizeof *req, "urlreq");
```

Comment:
This struct is used because it is used to communicate between the command version and bsd.c - we don't technically need the struct here, but it means that if there are edits for cmd.c, it's easier to tell what equivalent areas to edit.